Add CircleCI and related fixes

pauricthelodger opened the pull request:
This removes Shippable as the used CI and moves us over to CircleCI.

Shippable's runtime was causing issues and didn't easily allow testing across multiple OSes.
With Circle, we're able to test across multiple versions of Ubuntu, with different versions of GDAL, and different versions of python. Unfortunately, they didn't have a full matrix-type config available but I think I've kept it fairly clean.

I've set us with the the 3 current Ubuntu LTS releases: xenial (outgoing), bionic, focal (incoming)
I'm testing python2 only on xenial atm since that's on the way out but we can add it as a job on the others. The version of python3 is the default for each release so we get 3.5, 3.6, and 3.8.

GDAL is being tested with both the Ubuntu default and the ubuntugis PPA, stable (aka ppa) or unstable branches, experimental could be added. Focal just has the default as it's using the most recent version atm.

I added a fix / additional workaround for the pyvips issue, it was __slots__-related :grimacing: I've unpinned it in setup.py again but we should test this on heavier workloads to make sure we don't hit garbage collection issues before packaging a release.

I've updated the tests to allow the png8 renderer tests to have a different hash to compare against on each OS as it seems to change as versions go by. Focal looks like other tests have the same issue.

There's some additional small fixes and docs updates that came out of the pytest errors and warnings.

Currently the xenial default gdal tests are failing as there doesn't seem to be a PyPI version for 1.11.3. I could try dropping in the [pygdal](https://github.com/nextgis/pygdal) package to test there?

Bionic unstable gdal (3.0.4) is failing, as is Focal which is the same GDAL version. I think updating gdal2mbtiles for that is a separate task.

tomowatt commented on tests/test_renderers.py:
```diff
@@ -4,12 +4,37 @@
+PNG8_OS_HASHES = {
+    ('Ubuntu', '18.04', 'xenial'): 106831624867432276165545554861383631224,
+    ('Ubuntu', '18.04', 'bionic'): 93909651943814796643456367818041361877,
+    ('Ubuntu', '18.04', 'focal'): 226470660062402177473163372260043882022,
```
The OS version numbers need correcting

pauricthelodger replied:
Will do!

pauricthelodger commented:
> I'm not against this commit given the reasoning but there was a recent commit to pin `pyvips` to fix issues due to it being unpinned for others thinking
> 
> [45b00fc#diff-2eeaed663bd0d25b7e608891384b7298](https://github.com/ecometrica/gdal2mbtiles/commit/45b00fc0177855f89fffe1b3676d0eab9663abe5#diff-2eeaed663bd0d25b7e608891384b7298)

Yeah, I've explicitly set it to 2.1.8 for the minute for most envs in the tox.ini and unpinned again in setup.py since it was rolling back in the vips env after getting the most recent version. My aim is to make a `-pre` package build so we can test internally on larger data to see if garbage collection causes us issues.

I could add a note to README to say they should explicitly pin for the minute? But if someone is just automatically pip installing the head of `master` they'd still have an issue there.

pauricthelodger commented:
> As mentioned an additional test for `Xenial` with `PyVips` would be nice but everything else is looking good!

@tomowatt - done =)

pauricthelodger commented:
> As mentioned an additional test for `Xenial` with `PyVips` would be nice but everything else is looking good!

@tomowatt - done =)

pauricthelodger commented on tests/test_gdal.py:
```diff
@@ -254,9 +254,9 @@ def test_invalid_input(self):
                               outputfile=tmpfile.name)
 
-    @pytest.mark.skip_on_shippable
+    @pytest.mark.skip_on_ci
```
Ideally we'd either fix these so they work everywhere but I'm counting that as outside the scope of this ticket until CI tests actually work at all =)

pauricthelodger commented:
Updated README and changelog.
Fixed tzdata on Focal CI tests, now show the expected failure.
Flipped the vips tox env to test for the pinned version and that by default we now test unpinned.

jordiecometrica commented on tests/test_renderers.py:
```diff
@@ -4,12 +4,37 @@
+PNG8_OS_HASHES = {
+    ('Ubuntu', '16.04', 'xenial'): 106831624867432276165545554861383631224,
+    ('Ubuntu', '18.04', 'bionic'): 93909651943814796643456367818041361877,
+    ('Ubuntu', '20.04', 'focal'): 226470660062402177473163372260043882022,
```
Hm, how annoying the hash isn't stable. This will require more maintenance down the line.

pauricthelodger replied:
Very, but I'm putting it down to potential improvements in their renderer between versions, though I haven't measured file size ;)

jordiecometrica commented on README.rst:
```diff
@@ -74,10 +74,9 @@ You will need to install the PyPi GDAL package with the following options::
-    $ sudo apt-get install libvips libvips-dev
+Under Debian or Ubuntu, run the following to install it without the GUI nip2::
 
+    $ sudo apt-get install --no-install-recommends libvips libvips-dev
```
Huh, this is unusual, why no recommends?

pauricthelodger replied:
That's the "without the GUI nip2" bit. I seem to have have lost a comment when I fixup'd instead of squash =(
nip2 installs firefox and maybe some other big things that we don't need
